sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -4,18 +4,635 @@
+  }
+
+  // Allocate the instance of the response on the heap with a shared ptr so this memory gets
+  // delegated outside the transport and will be eventually released.
+  auto rawResponse
+      = std::make_unique<RawResponse>(majorVersion, minorVersion, httpStatusCode, reasonPhrase);
+
+  AddHeaders(responseHeaders, rawResponse);
+
+  int64_t contentLength
+      = GetContentLength(handleManager, requestMethod, rawResponse->GetStatusCode());
+
+  rawResponse->SetBodyStream(std::move(GetBodyStream(handleManager, contentLength)));
+  return rawResponse;
+}
+
+std::unique_ptr<RawResponse> WinHttpTransport::Send(Context const& context, Request& request)
+{
+  // TODO: Need to test if context has been canceled and cleanup, can't call ThrowIfCanceled.
+  (void)(context);
+
+  auto handleManager = std::make_unique<Details::HandleManager>(context, request);
+
+  GetSessionHandle(handleManager);
+  GetConnectionHandle(handleManager);
+  GetRequestHandle(handleManager);
+
+  SendRequest(handleManager);
+
+  ReceiveResponse(handleManager);
+
+  return GetRawResponse(handleManager, request.GetMethod());
+}
+
+// Read the response from the sent request.
+int64_t Details::WinHttpStream::Read(Context const& context, uint8_t* buffer, int64_t count)
+{
+  context.ThrowIfCanceled();
+
+  if (count <= 0 || this->m_isEOF)
+  {
+    return 0;
+  }
+
+  DWORD numberOfBytesRead = 0;
+
+  // Check for available data.
+  DWORD numberOfBytesAvailable = 0;
+  if (!WinHttpQueryDataAvailable(this->m_handleManager->m_requestHandle, &numberOfBytesAvailable))
+  {
+    // Errors include:
+    // ERROR_WINHTTP_CONNECTION_ERROR
+    // ERROR_WINHTTP_INCORRECT_HANDLE_STATE
+    // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_WINHTTP_OPERATION_CANCELLED
+    // ERROR_WINHTTP_TIMEOUT
+    // ERROR_NOT_ENOUGH_MEMORY
+
+    DWORD error = GetLastError();
+    throw Azure::Core::Http::TransportException(
+        "Error while querying how much data is available to read. Error Code: "
+        + std::to_string(error) + ".");
+  }
+
+  context.ThrowIfCanceled();
+
+  DWORD numberOfBytesToRead = numberOfBytesAvailable;
+  if (numberOfBytesAvailable > count)
+  {
+    numberOfBytesToRead = static_cast<DWORD>(count);
+  }
+
+  if (!WinHttpReadData(
+          this->m_handleManager->m_requestHandle,
+          (LPVOID)(buffer),
+          numberOfBytesToRead,
+          &numberOfBytesRead))
+  {
+    // Errors include:
+    // ERROR_WINHTTP_CONNECTION_ERROR
+    // ERROR_WINHTTP_INCORRECT_HANDLE_STATE
+    // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_WINHTTP_OPERATION_CANCELLED
+    // ERROR_WINHTTP_RESPONSE_DRAIN_OVERFLOW
+    // ERROR_WINHTTP_TIMEOUT
+    // ERROR_NOT_ENOUGH_MEMORY
+
+    DWORD error = GetLastError();
+    throw Azure::Core::Http::TransportException(
+        "Error while querying how much data is available to read. Error Code: "
```

Comment:
this error looks wrong. Should be error while reading?

Reply:
I don't understand what you mean. Can you clarify the question.

Reply:
Oh I get it, this is during `WinHttpReadData`. Copy/paste typo. Fixing.